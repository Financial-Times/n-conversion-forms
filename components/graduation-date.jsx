import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const getMonthNameByIndex = index => {
	return new Date(null, index, null).toLocaleDateString('en', { month: 'long' });
};

export function GraduationDate ({
	value,
	isDisabled = false,
	isRequired = false,
	hasError = false,
}) {
	const currentYear = (new Date()).getUTCFullYear();
	const earliestSelectableYear = currentYear - 4;
	const defaultValues = value && value.trim().match(/^([0-9]{4})-(0[0-9]|1[012])-[0-9]{2}$/);
	const defaultMonth = defaultValues && defaultValues[2];
	const defaultYear = defaultValues && defaultValues[1];
	const inputWrapperClassNames = classNames([
		'o-forms-input',
		'o-forms-input--select',
		{ 'o-forms-input--invalid': hasError },
	]);

	return (
		<label
			id="graduationDateField"
			className="o-forms-field ncf__validation-error"
			data-validate="required"
			htmlFor="graduationDateMonth"
		>
			<span className="o-forms-title">
				<span className="o-forms-title__main">Gradiation date</span>
			</span>
			<span className={inputWrapperClassNames}>
				<select
					id="graduationDateMonth"
					name="graduationDateMonth"
					required={isRequired}
					aria-required={isRequired}
					disabled={isDisabled}
					defaultValue={defaultMonth}
				>
					{Array.from({ length: 12 }, (_, index) => {
						return <option key={index} value={(index + 1).toString(10).padStart(2, '0')}>{getMonthNameByIndex(index + 1)}</option>;
					})}
				</select>
				<select
					id="graduationDateYear"
					name="graduationDateYear"
					required={isRequired}
					aria-required={isRequired}
					disabled={isDisabled}
					defaultValue={defaultYear}
				>
					{Array.from({ length: 9 }, (_, index) => {
						return <option key={index} value={earliestSelectableYear + index}>{earliestSelectableYear + index}</option>;
					})}
				</select>
				<span className="o-forms-input__error">Please select a valid graduation date</span>
			</span>
		</label>
	);
};

GraduationDate.propTypes = {
	value: PropTypes.string,
	isDisabled: PropTypes.bool,
	isRequired: PropTypes.bool,
	hasError: PropTypes.bool,
};
