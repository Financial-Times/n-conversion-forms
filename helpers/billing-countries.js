import { billingCountries } from 'n-common-static-data';

export { billingCountries };
