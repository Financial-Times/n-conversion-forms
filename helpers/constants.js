const printRegions = {
	cemeaV1: 'CEMEA_V1',
	cemeaV2: 'CEMEA_V2',
	apac: 'APAC',
};

const addressLine1AutoComplete = 'address-line1';
const addressLine2AutoComplete = 'address-line2';
const addressLine3AutoComplete = 'address-line3';

module.exports = {
	printRegions,
	addressLine1AutoComplete,
	addressLine2AutoComplete,
	addressLine3AutoComplete,
};
