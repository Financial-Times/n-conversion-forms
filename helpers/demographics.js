const { demographics } = require('n-common-static-data');

export { demographics };
