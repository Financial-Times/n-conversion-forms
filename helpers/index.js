module.exports = {
	'ncf-countries': require('./ncf-countries'),
	'ncf-common-data': require('./ncf-common-data'),
	cemeaV1ISO: require('./supportedCountries').cemeaV1ISO,
	cemeaV2ISO: require('./supportedCountries').cemeaV2ISO,
	apacISO: require('./supportedCountries').apacISO,
	identifyFTShippingZone: require('./supportedCountries')
		.identifyFTShippingZone,
	countriesSupported: require('./supportedCountries').countriesSupported,
	countriesSupportedISO: require('./supportedCountries').countriesSupportedISO,
	supportedPostcodeValidators: require('./supportedPostcodeValidators')
		.supportedPostcodeValidators,
	allSupportedPostcodeExamples: require('./supportedPostcodeExamples')
		.allSupportedPostcodeExamples,
	demographics: require('./demographics').demographics,
	billingCountries: require('./billing-countries').billingCountries,
};
