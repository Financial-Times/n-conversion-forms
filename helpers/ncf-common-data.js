const _get = require('lodash.get');
const COMMON_DATA = require('n-common-static-data');

module.exports = function ({ hash, fn }) {
	const context = Object.assign({}, { [hash.export]: _get(COMMON_DATA, hash.import) }, this);
	return fn(context);
};
