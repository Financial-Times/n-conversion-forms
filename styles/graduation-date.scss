@mixin ncfGraduationDate() {
	&__graduation-date {
		select {
			width: auto;
			margin-right: 12px;
			padding-right: 5em;
		}

		&__select-wrapper {
			display: flex;
		}
	}

	&__gd-compliance {
		white-space: pre-wrap;
	}
}
