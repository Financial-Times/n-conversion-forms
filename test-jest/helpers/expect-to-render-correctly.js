import ReactDOMServer from 'react-dom/server';

export const expectToRenderCorrectly = {
	toRenderCorrectly(reactComponent, props) {
		// Render the React component with props.
		const ReactComponentRendered = ReactDOMServer.renderToStaticMarkup(
			reactComponent(props)
		);

		// eslint-disable-next-line jest/no-standalone-expect
		expect(ReactComponentRendered).toMatchSnapshot();

		// This point is reached when the above assertions were successful.
		// The test should therefore always pass, that means it needs to be
		// `true` when used normally, and `false` when `.not` was used.
		return {
			pass: !this.isNot,
		};
	},
};
