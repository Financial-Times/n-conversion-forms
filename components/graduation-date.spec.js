import React from 'react';
import { GraduationDate, Compliance } from './graduation-date';
import { expectToRenderCorrectly } from '../test-jest/helpers/expect-to-render-correctly';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

expect.extend(expectToRenderCorrectly);
Enzyme.configure({ adapter: new Adapter() });

describe('GraduationDate', () => {
	it('renders with default props', () => {
		expect(GraduationDate).toRenderCorrectly({});
	});

	it('renders graduation date compliance component', () => {
		const wrapper = shallow(<GraduationDate />);
		expect(wrapper.find(Compliance).exists()).toBe(true);
		expect(
			wrapper
				.find(Compliance)
				.shallow()
				.find('#graduationDateCompliance')
				.exists()
		).toBe(true);
	});

	it('should display graduationDateMonth options as English month names', () => {
		const wrapper = shallow(<GraduationDate />);
		const values = wrapper
			.find('#graduationDateMonth option')
			.map((node) => node.text());
		expect(values).toEqual([
			'January',
			'February',
			'March',
			'April',
			'May',
			'June',
			'July',
			'August',
			'September',
			'October',
			'November',
			'December',
		]);
	});

	it('should store graduationDateMonth values as two-digit numbers', () => {
		const wrapper = shallow(<GraduationDate />);
		const values = wrapper
			.find('#graduationDateMonth option')
			.map((node) => node.prop('value'));
		expect(values).toEqual([
			'01',
			'02',
			'03',
			'04',
			'05',
			'06',
			'07',
			'08',
			'09',
			'10',
			'11',
			'12',
		]);
	});

	it('populates the graduationDateYear options with current year, +/- 4 years', () => {
		const stubDate = new Date('2020-08-26T00:00:00.000Z');
		jest.spyOn(global, 'Date').mockImplementation(() => stubDate);

		const wrapper = shallow(<GraduationDate />);
		const values = wrapper
			.find('#graduationDateYear option')
			.map((node) => node.prop('value'));

		expect(values).toEqual([
			2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024,
		]);

		jest.restoreAllMocks();
	});

	describe('given a valid value prop is passed', () => {
		const wrapper = shallow(<GraduationDate value="2020-08-28" />);

		it('should set the default month', () => {
			expect(wrapper.find('#graduationDateMonth').prop('defaultValue')).toEqual(
				'08'
			);
		});

		it('should set the default year', () => {
			expect(wrapper.find('#graduationDateYear').prop('defaultValue')).toEqual(
				'2020'
			);
		});
	});

	describe('given an invalid value prop is passed', () => {
		const wrapper = shallow(<GraduationDate value="invalid" />);

		it('should not set any default values', () => {
			expect(
				wrapper.find('#graduationDateMonth').prop('defaultValue')
			).toBeFalsy();
			expect(
				wrapper.find('#graduationDateYear').prop('defaultValue')
			).toBeFalsy();
		});
	});

	describe('given the isDisabled prop is set to true', () => {
		const wrapper = shallow(<GraduationDate isDisabled={true} />);

		it('should disable both select fields', () => {
			expect(
				wrapper.find('#graduationDateMonth').prop('disabled')
			).toBeTruthy();
			expect(wrapper.find('#graduationDateYear').prop('disabled')).toBeTruthy();
		});
	});

	describe('given the isRequired prop is set to true', () => {
		const wrapper = shallow(<GraduationDate isRequired={true} />);

		it('should make both fields required', () => {
			expect(
				wrapper.find('#graduationDateMonth').prop('required')
			).toBeTruthy();
			expect(wrapper.find('#graduationDateYear').prop('required')).toBeTruthy();
		});
	});

	describe('given the hasError prop is set to true', () => {
		const wrapper = shallow(<GraduationDate hasError={true} />);

		it('should add an invalid modifier class', () => {
			expect(wrapper.find('div.o-forms-input').prop('className')).toMatch(
				'o-forms-input--invalid'
			);
		});
	});
});
