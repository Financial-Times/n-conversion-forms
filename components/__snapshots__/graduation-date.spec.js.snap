// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraduationDate renders with default props 1`] = `
<label id="graduationDateField"
       class="o-forms-field ncf__validation-error ncf__graduation-date"
       data-validate="required"
       for="graduationDateMonth"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Graduation date
    </span>
  </span>
  <span class="o-forms-input o-forms-input--select ncf__graduation-date__select-wrapper">
    <span>
      <label for="graduationDateMonth">
        Month
      </label>
      <select id="graduationDateMonth"
              name="graduationDateMonth"
              aria-required="false"
      >
        <option value="01">
          January
        </option>
        <option value="02">
          February
        </option>
        <option value="03">
          March
        </option>
        <option value="04">
          April
        </option>
        <option value="05">
          May
        </option>
        <option value="06">
          June
        </option>
        <option value="07">
          July
        </option>
        <option value="08">
          August
        </option>
        <option value="09">
          September
        </option>
        <option value="10">
          October
        </option>
        <option value="11">
          November
        </option>
        <option value="12">
          December
        </option>
      </select>
    </span>
    <span>
      <label for="graduationDateYear">
        Year
      </label>
      <select id="graduationDateYear"
              name="graduationDateYear"
              aria-required="false"
      >
        <option value="2016">
          2016
        </option>
        <option value="2017">
          2017
        </option>
        <option value="2018">
          2018
        </option>
        <option value="2019">
          2019
        </option>
        <option value="2020">
          2020
        </option>
        <option value="2021">
          2021
        </option>
        <option value="2022">
          2022
        </option>
        <option value="2023">
          2023
        </option>
        <option value="2024">
          2024
        </option>
      </select>
    </span>
    <span class="o-forms-input__error">
      Please select a valid graduation date
    </span>
  </span>
</label>
`;
